### PR TITLE
Add more test cases for graph API + fix bug

### PR DESCRIPTION
**Pull request by jacobperron**
Guard against invalid allocator in rcl_get_service_names_and_types.

### CODE REVIEW

**Comment by jacobperron** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/6462/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/2832/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/5307/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/6281/)

**Comment by jacobperron** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/6469/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/2836/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/5311/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/6285/)

**Comment by jacobperron** [comment]
Rebased on master:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/6583/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/2921/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/5399/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/6393/)

File: rcl/test/rcl/test_graph.cpp [hunk 1]
```diff
@@ -205,9 +246,186 @@ TEST_F(
...
+    &zero_node, &allocator, false, node_name, node_ns, &nat);
+  EXPECT_EQ(RCL_RET_NODE_INVALID, ret) << rcl_get_error_string().str;
+  rcl_reset_error();
+  ret = rcl_get_publisher_names_and_types_by_node(
+    this->old_node_ptr, &allocator, false, node_name, node_ns, &nat);
+  EXPECT_EQ(RCL_RET_NODE_INVALID, ret) << rcl_get_error_string().str;
+  rcl_reset_error();
+  // invalid allocator
+  ret = rcl_get_publisher_names_and_types_by_node(
+    this->node_ptr, nullptr, false, node_name, node_ns, &nat);
```
Comment (hidmic) [inline rcl/test/rcl/test_graph.cpp]: @jacobperron here and elsewhere, consider also testing with a non-null but invalid allocator e.g. a zero initialized one as returned by [`rcutils_get_zero_initialized_allocator()`](https://github.com/ros2/rcutils/blob/master/include/rcutils/allocator.h#L84).

File: rcl/test/rcl/test_graph.cpp [hunk 2]
```diff
@@ -205,9 +246,186 @@ TEST_F(
...
+  // invalid names
+  ret = rcl_get_publisher_names_and_types_by_node(
+    this->node_ptr, &allocator, false, nullptr, node_ns, &nat);
+  EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, ret) << rcl_get_error_string().str;
+  rcl_reset_error();
+  ret = rcl_get_publisher_names_and_types_by_node(
+    this->node_ptr, &allocator, false, node_name, nullptr, &nat);
+  EXPECT_EQ(RCL_RET_INVALID_ARGUMENT, ret) << rcl_get_error_string().str;
+  rcl_reset_error();
+  // TODO(jacobperron): test valid strings with invalid topic names in them
```
[the remark this answers is not in the document]
Comment (jacobperron) [reply rcl/test/rcl/test_graph.cpp]: Done for the added tests. 5a8597c

File: rcl/test/rcl/test_graph.cpp (hunk 1 above)
Comment (jacobperron) [reply rcl/test/rcl/test_graph.cpp]: Done 039e5fe

**Comment by jacobperron** [comment]
@hidmic Thanks for the review! In refactoring the tests, I've discovered an issue were the rmw layer doesn't seem to validate namespaces like it does for node names. I've added some TODOs in the test and I'll open an issue to look into this separately.

**Comment by jacobperron** [comment]
Opened an issue https://github.com/ros2/rcl/issues/414

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/6654/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/2978/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/5465/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/6485/)
